File: Leetcode - Python/1013_array_canThreePartsEqualSum.py
# https://leetcode.com/problems/partition-array-into-three-parts-with-equal-sum/

class Solution:
    def canThreePartsEqualSum(self, arr: List[int]) -> bool:
        l, r, s = 1, len(arr) - 2, sum(arr)
        leftsum, rightsum, average = arr[0], arr[-1], s // 3

        while l <= r:
            if l < r and leftsum != average:
                leftsum += arr[l]
                l += 1

            if l < r and rightsum != average:
                rightsum += arr[r]
                r -= 1

            if leftsum == average and rightsum == average and s % 3 == 0:
                return True

            if l == r:
                break

        return False

File: Leetcode - Java/1013_array_canThreePartsEqualSum.java
// https://leetcode.com/problems/partition-array-into-three-parts-with-equal-sum/

class Solution {
    public boolean canThreePartsEqualSum(int[] arr) {
        int sum = Arrays.stream(arr).sum(), average = sum / 3;
        int l = 0, r = arr.length - 1;
        int leftsum = arr[l++], rightsum = arr[r--];
        
        do {
            if (leftsum != average) {
                leftsum += arr[l++];
            }
            if (l < r && rightsum != average) {
                rightsum += arr[r--];
            }
            if (l <= r && sum % 3 == 0 && leftsum == average && rightsum == average) {
                return true;
            } 
        } while (l < r);
        return false;
    }
}